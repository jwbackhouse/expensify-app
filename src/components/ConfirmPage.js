import React from 'react';
import { Link } from 'react-router-dom';
import {firebase} from '../firebase/firebase';
import Header from './Header';


export class ConfirmPage extends React.Component {
  state = {
    email: '',
    error: '',
    showForm: false
  };
  
  componentDidMount = () => {
    this.isValidURL();
  }
  
  onEmailChange = (e) => {
    const email = e.target.value;
    this.setState(() => ({ email }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    this.signIn(this.state.email);
  };
  
  signIn = (email) => {
    firebase.auth().signInWithEmailLink(email, window.location.href)    // auth code parsed from URL
      .then((result) => {
        window.localStorage.removeItem('emailForSignIn');   // Clear email from storage.
        // Can access the new user via result.user
      })
      .catch((err) => {
        const error = 'Sorry, something went wrong.'
        this.setState(() => ({error}))
      });
  };
  
  isValidURL = () => {
    const pageURL = firebase.auth().isSignInWithEmailLink(window.location.href);
    const savedEmail = window.localStorage.getItem('emailForSignIn');
    if (pageURL) {    // Check page URL contains valid auth code
      if (!savedEmail) {    // If local storage doesn't contain an email (e.g. user switched devices)
        const showForm = true;
        this.setState(() => ({showForm}))
      } else {
        this.signIn(savedEmail);
      }
    } else {
      const error = "Sorry, that doesn't look like a valid link."
      this.setState(() => ({error}))
    }
    
  }
  
  onClick = () => {
    this.isValidURL();
  }

  render() {
    let form, error;
    if (this.state.showForm) {
      form = (
        <div>
          <h3>Please confirm your email</h3>
          <form onSubmit = { this.onSubmit }>
            <input
              type='email'
              placeholder='Email address'
              value={this.state.email}
              onChange={this.onEmailChange}
            />
            <button className="button--tertiary-pink">Confirm</button>
          </form>
        </div>
      )
    }
    if (this.state.error) {
      error = (
        <div>
          <p>{ this.state.error }</p>
          <Link className='button' to='/'>Return to login page</Link>
        </div>
      )
    }
    return (
      <div className='box-layout'>
        <div className='box-layout__box'>
          <h1 className='box-layout__title'>Expensify</h1>
          { form }
          { error }
        </div>
      </div>
    )
  }
};

export default ConfirmPage;