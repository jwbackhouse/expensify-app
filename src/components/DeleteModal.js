import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';

// TODO:
export class DeleteModal extends React.Component {
  componentDidMount = () => {
    Modal.setAppElement('body');
  }
  
  render() {
    return (
      <Modal
        isOpen={!!this.props.showModal}
        onRequestClose={this.props.onModalCancel}   // Allows esc / clicking background to close modal
        contentLabel='Just to check...'
        closeTimeoutMS={ 200 }    // Allows us to target the before-close class to apply a transition
        className='modal'
      >
        <h3 className='modal__title'>Are you sure you want to delete this expense?</h3>
        <button
          className='button button--green button--margin-right'
          id='confirm'
          onClick={this.props.onModalConfirm}
        >Yes</button>
        <button
          className='button button--grey'
          onClick={ this.props.onModalCancel }
          id='cancel'
        >No</button>
      </Modal>
    );
  }
};

export default DeleteModal;