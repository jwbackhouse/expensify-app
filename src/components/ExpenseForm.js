import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import NumberFormat from 'react-number-format';

export default class ExpenseForm extends React.Component {
  // Populate fields if data passed on EditPage, otherwise leave blank for AddExpensePage
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    note: this.props.expense ? this.props.expense.note : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: ''
  };
  
  // Update state.description with typed value, which then renders into the form
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  
  // Ditto for note
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  
  // Ditto for amount
  onAmountChange = (values) => {
    const amount = values.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {    // Allow no entry || Enforce max two decimal places
      this.setState(() => ({amount}));
    }
  };
  
  // Date picker functions
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({createdAt}));
    }   // No else clause so cannot delete date field
  };
  
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  
  onSubmit = (e) => {
    e.preventDefault();
    
    if (!this.state.description || !this.state.amount) {    // Check description and amount are present
      this.setState(() => ({error: 'Please complete both the description and amount'}));    // Error message
    } else {
      this.setState(() => ({error: ''}));   // Clear error message
      this.props.onSubmit({   // Provide expense object to props function
        description: this.state.description,
        amount: parseFloat(this.state.amount,10) * 100,    // Convert amount string to number and convert into pence
        createdAt: this.state.createdAt.valueOf(),   // Convert moment object to Unix ms
        note: this.state.note
      });
    }
  };
  
  // Render
  render() {
    return (
      <div>
        <form className='form' onSubmit = { this.onSubmit }>
          { this.state.error && <p className='form__error'>{this.state.error}</p> }
          <input
            type = 'text'
            className='text-input'
            placeholder = 'Description'
            value = {this.state.description}
            onChange = {this.onDescriptionChange}
            autoFocus
          />
          <NumberFormat
            value={ this.state.amount }
            className='text-input'
            id='amount-input'
            onValueChange = { this.onAmountChange }
            placeholder = 'Amount'
            // Formatting options
            allowEmptyFormatting={false}
            decimalScale={2}
            fixedDecimalScale={false}   // Set to true to always show decimal places
            isNumericString={ true }    // Needed to accept state values
            prefix={'Â£'}
            thousandSeparator={true}
          />
          <SingleDatePicker
            date = { this.state.createdAt }
            onDateChange = { this.onDateChange }
            focused = { this.state.calendarFocused }    // Is the calendar open or closed
            onFocusChange = { this.onFocusChange }
            numberOfMonths = { 1 }    // # months to show
            isOutsideRange={() => false }   // Allow dates in the past
            displayFormat={() => "DD/MM/YYYY"}    // Convert to UK date format
          />
          <textarea
            type = 'text'
            className='textarea'
            value = {this.state.note}
            placeholder = 'Enter details here'
            onChange = { this.onNoteChange }
          ></textarea>
          <div>
            <button className='button button--green'>
              { this.props.buttonText ? this.props.buttonText : 'Add expense' }
            </button>
          </div>
        </form>
      </div>
    )
  }
};