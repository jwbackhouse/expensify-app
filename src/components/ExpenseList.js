import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import visibleExpenses from '../selectors/expenses.js';
import {
  sortByAmountAsc,
  sortByAmountDesc,
  sortByDateNewest,
  sortByDateOldest,
  sortByNameAsc,
  sortByNameDesc
} from '../actions/filters';

// Render an ExpenseListItem component for each expense
export const ExpenseList = (props) => (
  <div className='content-container'>
    <div className='list-header'>
      <div className='show-for-mobile'>
        Expenses
      </div>
      <div className='show-for-desktop show-for-desktop--primary' id='sortByName'>
        Expense
        <img
          className='arrow arrow--up'
          onClick = {() =>{ props.sortByNameAsc() }}
          src='/images/arrow-up.jpg'
        />
        <img
          className='arrow arrow--down'
          onClick = {() => {props.sortByNameDesc()}}
          src='/images/arrow-down.jpg'
        />
      </div>
      <div className='show-for-desktop' id='sortByDate'>
        Date created
        <img
          className='arrow arrow--up'
          onClick = {() =>{ props.sortByDateOldest() }}
          src='/images/arrow-up.jpg'
        />
        <img
          className='arrow arrow--down'
          onClick = {() => {props.sortByDateNewest()}}
          src='/images/arrow-down.jpg'
        />
      </div>
      <div className='show-for-desktop' id='sortByAmount'>
        Amount
        <img
          className='arrow arrow--up'
          onClick = {() =>{ props.sortByAmountAsc() }}
          src='/images/arrow-up.jpg'
        />
        <img
          className='arrow arrow--down'
          onClick = {() => {props.sortByAmountDesc()}}
          src='/images/arrow-down.jpg'
        />
      </div>
    </div>
    <div className='list-body'>
      { props.expenses.length === 0
        ? <div className='list-item list-item--message'>
            <span>No expenses yet</span>
          </div>
        : props.expenses.map((expense) => {
            return <ExpenseListItem key = { expense.id } { ...expense }/>
          })
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: visibleExpenses(state.expenses, state.filters),    // or to return all expenses use state.expenses
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    sortByDateNewest: () => dispatch(sortByDateNewest()),
    sortByDateOldest: () => dispatch(sortByDateOldest()),
    sortByAmountAsc: () => dispatch(sortByAmountAsc()),
    sortByAmountDesc: () => dispatch(sortByAmountDesc()),
    sortByNameAsc: () => dispatch(sortByNameAsc()),
    sortByNameDesc: () => dispatch(sortByNameDesc()),
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);