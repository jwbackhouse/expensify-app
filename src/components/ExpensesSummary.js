import React from 'react';
import { connect } from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import visibleExpenses from '../selectors/expenses.js';
import numeral from 'numeral';
// Set GB currency
import "numeral/locales/en-gb";
numeral.locale('en-gb');

export const ExpensesSummary = (props) => {
  const total = expensesTotal(props.expenses);
  const formattedTotal = numeral(total / 100).format('$0,0.00')
  const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses'
  return (
    <div>
      <p>Viewing { props.expenses.length } { expenseWord } totalling { formattedTotal }.</p>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: visibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);