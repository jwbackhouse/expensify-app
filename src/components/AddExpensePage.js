import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);    // Add expense to state under user's account
    this.props.history.push('/');    // Redirect on submit (uses in-built method)
  };
  
  render() {
    return (
      <div>
        <p>This is the add expense page</p>
        <ExpenseForm
          onSubmit={ this.onSubmit }
        />
      </div>
    );
  };
};

// Create bespoke dispatch method in order to simplify the above (and make it possible to test with a spy)
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);    // first arg is state, second is dispatch


// *** Previous in-line version of the above (which couldn't be tested in the same way)
// const AddExpensePage = (props) => (
//   <div>
//     <p>This is the add expense page</p>
//     <ExpenseForm onSubmit={(expense) => {
//         props.dispatch(addExpense(expense));    // Add expense to state
//         props.history.push('/');    // Redirect on submit (uses in-built method)
//       }}
//     />
//   </div>
// );