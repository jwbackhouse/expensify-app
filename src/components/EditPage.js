import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import DeleteModal from './DeleteModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditPage extends React.Component {
  state = {
    showModal: false
  };
  
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);    // Edit expense
    this.props.history.push('/');   // Redirect on submit (uses in-built method)
  };
  
  onClickDelete = () => {
    this.setState(() => ({ showModal: true }));
  };
  
  onClickCancel = () => {
    this.props.history.push('/');
  };
  
  onModalConfirm = () => {
    this.props.startRemoveExpense(this.props.expense.id);    // Delete expense
    this.setState(() => ({showModal: false}));
    this.props.history.push('/');   // Redirect on submit (uses in-built method)
  };
  
  onModalCancel = () => {
    this.setState(() => ({showModal: false}));
  };
  
  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Edit expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            expense = {this.props.expense}
            buttonText = { 'Save changes' }
            onSubmit = { this.onSubmit }
          />
          <div>
            <button
              className='button button--pink button--margin-right'
              id='deleteButton'
              onClick = { this.onClickDelete }
            >Delete</button>
            <button
              className='button button--grey'
              id='cancelButton'
              onClick = { this.onClickCancel }
            >Cancel</button>
          </div>
        </div>
        <DeleteModal onModalConfirm={ this.onModalConfirm } onModalCancel={ this.onModalCancel } showModal={this.state.showModal} />
      </div>
    );
  };
  
  // *** PREVIOUS CODE BEFORE MAPTODISPATCH ADDED
  // return (
  //   <div>
  //     <ExpenseForm
  //       expense = {props.expense}
  //       buttonText = { 'Save changes' }
  //       onSubmit = {(expense) => {
  //         props.dispatch(editExpense(props.expense.id, expense));    // Edit expense
  //         props.history.push('/');    // Redirect on submit (uses in-built method)
  //       }}
  //     />
  //     <button onClick = {() => {
  //       props.dispatch(removeExpense({ id: props.expense.id }));    // Delete expense
  //       props.history.push('/');
  //     }}>Delete</button>
  //   </div>
  // );
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
})

const mapStateToProps= (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)   // Matches ID pulled from URL
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);