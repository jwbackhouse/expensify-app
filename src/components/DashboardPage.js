import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilter from './ExpenseListFilter'

const DashboardPage = () => (
  <div>
    <p>This is the homepage</p>
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default DashboardPage;