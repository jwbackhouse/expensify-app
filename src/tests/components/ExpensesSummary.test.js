import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';   // import unconnected version so we can populate props
import testExpenses from '../fixtures/testExpenses';

test('Should render correct message for single expense with two hidden', () => {
  const wrapper = shallow(<ExpensesSummary visibleExpenses={[testExpenses[0]]} allExpenses={ testExpenses }/>);
  expect(wrapper).toMatchSnapshot();
})

test('Should render correct message for multiple expenses with none hidden', () => {
  const wrapper = shallow(<ExpensesSummary visibleExpenses={ testExpenses } allExpenses={ testExpenses }  />);
  expect(wrapper).toMatchSnapshot();
})