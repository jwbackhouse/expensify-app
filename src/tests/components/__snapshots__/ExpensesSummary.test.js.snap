// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correct message for multiple expenses with none hidden 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        £3.10
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <[object Object]
        className="button button--pink"
        to="/create"
      >
        Add expense
      </[object Object]>
    </div>
  </div>
</div>
`;

exports[`Should render correct message for single expense with two hidden 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        £1.00
      </span>
    </h1>
    <p
      className="page-header__subtitle"
    >
      (
      2
       hidden 
      expenses
       totalling 
      £2.10
       not showing)
    </p>
    <div
      className="page-header__actions"
    >
      <[object Object]
        className="button button--pink"
        to="/create"
      >
        Add expense
      </[object Object]>
    </div>
  </div>
</div>
`;
